### PR TITLE
Error Handling Added

### PR DESCRIPTION
**Pull request by lalib**
The clients can use their own handlers as many as they want.

They should just implement the EventErrorHandler interface.

By default, there is no handler.

### CODE REVIEW

File: paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/partitioned/impl/SimplePartitionCoordinator.java [hunk 1]
```diff
@@ -13,16 +15,24 @@
...
 
 public class SimplePartitionCoordinator extends AbstractPartitionCoordinator {
 
     private final AtomicBoolean running = new AtomicBoolean(true);
     private boolean startNewestAvailableOffset = true;
+    private final List<EventErrorHandler> eventErrorHandlerList;
 
     public SimplePartitionCoordinator() {
         super(LoggerFactory.getLogger(SimplePartitionCoordinator.class));
+        eventErrorHandlerList = Collections.emptyList();
```
Comment (GJL) [inline paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/partitioned/impl/SimplePartitionCoordinator.java]: [C] Code duplication can be avoided by calling this(Collections.emptyList());

File: paradox-nakadi-consumer-partitioned-zk/src/main/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagement.java [hunk 2]
```diff
@@ -58,14 +67,19 @@ public void flush(final EventTypePartition eventTypePartition) {
...
         if (ThrowableUtils.isUnrecoverableException(t)) {
             LOGGER.error("Error [{}] reason [{}]", eventTypePartition, getMessage(t));
             ThrowableUtils.throwException(t);
         } else {
-            LOGGER.error("Error [{}] reason [{}]", eventTypePartition, getMessage(t), t);
+
+            eventErrorHandlers.forEach(eventErrorHandler ->
+                    eventErrorHandler.onError(t, eventTypePartition, offset, rawEvent));
+
+            LOGGER.error("Error [{}] reason [{}] raw event [{}] ", eventTypePartition, getMessage(t), t, rawEvent);
```
Comment (GJL) [inline paradox-nakadi-consumer-partitioned-zk/src/main/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagement.java]: [C] `t` should be the last argument. Also see http://slf4j.org/faq.html#paramException

File: paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import org.junit.Test;
+
+import org.mockito.Mockito;
+
+import de.zalando.paradox.nakadi.consumer.core.domain.EventTypePartition;
+import de.zalando.paradox.nakadi.consumer.core.exceptions.UnrecoverableException;
+import de.zalando.paradox.nakadi.consumer.core.partitioned.PartitionCommitCallbackProvider;
+import de.zalando.paradox.nakadi.consumer.core.partitioned.PartitionRebalanceListenerProvider;
+
+public class ZKConsumerSyncOffsetManagementTest {
```
Comment (GJL) [inline paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java]: [C] Test with non-empty handler list is missing.

**Comment by lalib** [comment]
@GJL I've made the changes.

File: paradox-nakadi-consumer-boot/src/main/java/de/zalando/paradox/nakadi/consumer/boot/NakadiEventErrorHandlerConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+import org.zalando.stups.tokens.config.AccessTokensBeanAutoConfiguration;
+
+import de.zalando.paradox.nakadi.consumer.boot.components.EventErrorHandlerList;
+import de.zalando.paradox.nakadi.consumer.core.http.handlers.EventErrorHandler;
+
+@Configuration
+@AutoConfigureAfter(AccessTokensBeanAutoConfiguration.class)
```
Comment (GJL) [inline paradox-nakadi-consumer-boot/src/main/java/de/zalando/paradox/nakadi/consumer/boot/NakadiEventErrorHandlerConfiguration.java]: I think this is not required.

File: paradox-nakadi-consumer-boot/src/main/java/de/zalando/paradox/nakadi/consumer/boot/ReplayHandler.java [hunk 5]
```diff
@@ -31,21 +33,25 @@
 
     private static final PartitionCoordinator THROWING_COORDINATOR = new EmptyPartitionCoordinator() {
         @Override
-        public void error(final Throwable t, final EventTypePartition eventTypePartition) {
+        public void error(final Throwable t, final EventTypePartition eventTypePartition, @Nullable final String cursor,
+                final String rawEvent) {
+
+            // todo should we add the handler?
```
Comment (GJL) [inline paradox-nakadi-consumer-boot/src/main/java/de/zalando/paradox/nakadi/consumer/boot/ReplayHandler.java]: Error handler should be only added if the exception is not handled elsewhere.

File: paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/EventErrorHandler.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package de.zalando.paradox.nakadi.consumer.core.http.handlers;
+
+import javax.annotation.Nullable;
+
+import de.zalando.paradox.nakadi.consumer.core.domain.EventTypePartition;
+
+@FunctionalInterface
```
Comment (GJL) [inline paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/EventErrorHandler.java]: Nice

File: paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/RawEventResponseHandler.java [hunk 7]
```diff
@@ -16,7 +14,7 @@ public RawEventResponseHandler(final EventTypePartition eventTypePartition, fina
     }
 
     @Override
-    Optional<NakadiEventBatch<String>> getEventBatch(final String string) {
-        return EventUtils.getRawEventBatch(jsonMapper, string);
+    NakadiEventBatch<String> getEventBatch(final String string) {
+        return EventUtils.getRawEventBatch(jsonMapper, string).orElseThrow(IllegalArgumentException::new);
```
Comment (GJL) [inline paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/RawEventResponseHandler.java]: Original exception should be preserved, imho

File: paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/AbstractEventsResponseBulkHandler.java [hunk 8]
```diff
@@ -37,26 +37,26 @@ public void onResponse(final String content) {
...
-    private void handleEvents(final EventTypeCursor cursor, final List<T> events) {
+    private void handleEvents(final EventTypeCursor cursor, final List<T> events, final String content) {
         try {
             delegate.onEvent(cursor, events);
             coordinator.commit(cursor);
-        } catch (Throwable t) {
+        } catch (final Throwable t) {
             log.error("Handler error at cursor [{}]", cursor);
-            coordinator.error(t, eventTypePartition);
+            coordinator.error(t, eventTypePartition, cursor.getOffset(), content);
```
Comment (GJL) [inline paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/AbstractEventsResponseBulkHandler.java]: [C] `AbstractEventsResponseBulkHandler.getEventBatch()` still returns `Optional<NakadiEventBatch<T>>` but the exception should be propagated.

File: paradox-nakadi-consumer-partitioned-zk/src/main/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagement.java [hunk 9]
```diff
@@ -26,13 +32,16 @@
...
 
     ZKConsumerSyncOffsetManagement(final ZKConsumerOffset consumerOffset,
             final PartitionCommitCallbackProvider commitCallbackProvider,
-            final PartitionRebalanceListenerProvider rebalanceListenerProvider) {
+            final PartitionRebalanceListenerProvider rebalanceListenerProvider,
+            final List<EventErrorHandler> eventErrorHandlers) {
         this.commitCallbackProvider = commitCallbackProvider;
         this.rebalanceListenerProvider = rebalanceListenerProvider;
         this.consumerOffset = consumerOffset;
+        this.eventErrorHandlers = requireNonNull(eventErrorHandlers);
```
Comment (GJL) [inline paradox-nakadi-consumer-partitioned-zk/src/main/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagement.java]: I believe all the other parameters are also `@NonNull`

File: paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+import de.zalando.paradox.nakadi.consumer.core.domain.EventTypePartition;
+import de.zalando.paradox.nakadi.consumer.core.exceptions.UnrecoverableException;
+import de.zalando.paradox.nakadi.consumer.core.http.handlers.EventErrorHandler;
+import de.zalando.paradox.nakadi.consumer.core.partitioned.PartitionCommitCallbackProvider;
+import de.zalando.paradox.nakadi.consumer.core.partitioned.PartitionRebalanceListenerProvider;
+
+public class ZKConsumerSyncOffsetManagementTest {
+
+    @Test
+    public void testShouldNotThrowExceptionWhenThereIsNoHandler() {
```
Comment (GJL) [inline paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java]: Is this needed? There is no assert.

File: paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+        new ZKConsumerSyncOffsetManagement(Mockito.mock(ZKConsumerOffset.class),
+            Mockito.mock(PartitionCommitCallbackProvider.class), Mockito.mock(PartitionRebalanceListenerProvider.class),
+            Collections.emptyList()).error(new IllegalStateException(), Mockito.mock(EventTypePartition.class), "2324",
+            "event");
+    }
+
+    @Test
+    public void testShouldThrowExceptionWhenItGetsTheUnrecoverableException() {
+        final ZKConsumerSyncOffsetManagement zkConsumerSyncOffsetManagement = new ZKConsumerSyncOffsetManagement(Mockito
```
Comment (GJL) [inline paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java]: You don't like `@Mock` annotation?

File: paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        final EventErrorHandler spy = Mockito.spy(EventErrorHandler.class);
+
+        final ZKConsumerSyncOffsetManagement zkConsumerSyncOffsetManagement = new ZKConsumerSyncOffsetManagement(Mockito
+                    .mock(ZKConsumerOffset.class), Mockito.mock(PartitionCommitCallbackProvider.class),
+                Mockito.mock(PartitionRebalanceListenerProvider.class), Collections.singletonList(spy));
+
+        zkConsumerSyncOffsetManagement.error(new RuntimeException(), Mockito.mock(EventTypePartition.class), "2324",
+            "event");
+
+        Mockito.verify(spy).onError(Mockito.any(Throwable.class), Mockito.any(EventTypePartition.class),
```
Comment (GJL) [inline paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java]: This assert is very weak. Argument values should be verified. I prefer `ArgumentCaptor` + `assertThat`

File: paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        Assertions.assertThatThrownBy(() ->
+                          zkConsumerSyncOffsetManagement.error(new UnrecoverableException(),
+                              Mockito.mock(EventTypePartition.class), "2324", "event")).isInstanceOf(
+                      UnrecoverableException.class);
+    }
+
+    @Test
+    public void testShouldLogTheExceptionWithHandler() {
+
+        final EventErrorHandler spy = Mockito.spy(EventErrorHandler.class);
```
Comment (GJL) [inline paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java]: [C] Spy is not needed.

```
Creates a spy of the real object. The spy calls real methods unless they are stubbed.
Real spies should be used carefully and occasionally, for example when dealing with legacy code.
```

Here a normal mock object is used.

File: paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/JsonEventResponseHandler.java [hunk 14]
```diff
@@ -17,7 +15,7 @@ public JsonEventResponseHandler(final EventTypePartition eventTypePartition, fin
     }
 
     @Override
-    Optional<NakadiEventBatch<JsonNode>> getEventBatch(final String string) {
-        return EventUtils.getJsonEventBatch(jsonMapper, string);
+    NakadiEventBatch<JsonNode> getEventBatch(final String string) {
+        return EventUtils.getJsonEventBatch(jsonMapper, string).orElseThrow(IllegalStateException::new);
```
Comment (GJL) [inline paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/JsonEventResponseHandler.java]: [C] EventUtils should be probably changed into not returning `Optional`

File: gradle.properties [hunk 15]
```diff
@@ -1 +1 @@
-version=0.0.23-RC1
+version=0.1.0-beta1
```
Comment (GJL) [inline gradle.properties]: Can you revert this? I think it is better if we raise the version in a separate PR.

**Comment by lalib** [comment]
@GJL Could you review it again please?

File: paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/AbstractEventsResponseBulkHandler.java [hunk 16]
```diff
@@ -26,25 +25,20 @@
...
         final String[] events = getEvents(content);
-        for (String event : events) {
-            final Optional<NakadiEventBatch<T>> optionalBatch = getEventBatch(event);
-            if (!optionalBatch.isPresent()) {
-                return;
-            }
+        for (final String event : events) {
+            final NakadiEventBatch<T> nakadiEventBatch = getEventBatch(event);
+
+            if (nakadiEventBatch != null) {
```
Comment (GJL) [inline paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/AbstractEventsResponseBulkHandler.java]: [C] From what I have seen `nakadiEventBatch` cannot be null. How about `requireNonNull(getEventBatch(event))`?

File: paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/RawContentResponseHandler.java [hunk 17]
```diff
@@ -31,29 +29,29 @@ public RawContentResponseHandler(final EventTypePartition eventTypePartition, fi
     public void onResponse(final String content) {
         final String[] events = getEvents(content);
         if (events.length > 0) {
-            final Optional<NakadiEventCursor> optionalFirstCursor = getEventCursor(events[0]);
-            if (!optionalFirstCursor.isPresent()) {
-                return;
-            }
+            final NakadiEventCursor nakadiFirstCursor = getEventCursor(events[0]);
```
Comment (GJL) [inline paradox-nakadi-consumer-core/src/main/java/de/zalando/paradox/nakadi/consumer/core/http/handlers/RawContentResponseHandler.java]:
```
    NakadiEventCursor getEventCursor(final String string) {
        try {
            return jsonMapper.readValue(string, NakadiEventCursor.class);
        } catch (IOException e) {
            log.error("Error while parsing event cursor from [{}]", string, e);
            ThrowableUtils.throwException(e);
            return null;
        }
    }
```

[C] `getEventCursor()` cannot return null. Hence, the null check is not needed.

File: paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java [hunk 18]
```diff
@@ -16,40 +21,44 @@
...
-                Mockito.mock(PartitionRebalanceListenerProvider.class), Collections.singletonList(spy));
-
-        zkConsumerSyncOffsetManagement.error(new RuntimeException(), Mockito.mock(EventTypePartition.class), "2324",
-            "event");
+        zKConsumerSyncOffsetManagement.error(new RuntimeException(), eventTypePartition, "2324", "event");
 
-        Mockito.verify(spy).onError(Mockito.any(Throwable.class), Mockito.any(EventTypePartition.class),
-            Mockito.anyString(), Mockito.anyString());
+        Mockito.verify(eventErrorHandler).onError(Mockito.any(Throwable.class), Mockito.any(EventTypePartition.class),
+            eq("2324"), Mockito.anyString());
```
Comment (GJL) [inline paradox-nakadi-consumer-partitioned-zk/src/test/java/de/zalando/paradox/nakadi/consumer/partitioned/zk/ZKConsumerSyncOffsetManagementTest.java]: `eq("event")`

**Comment by lalib** [comment]
@GJL Could you please merge it?

**Pull request merged**